src/transformations/defmodel.hpp

```diff
@@ -0,0 +1,1890 @@
+            dZ01 = dn01 * cosphi01;
+
+            const double sinlam10 = sinhalfresx;
+            const double coslam10 = coshalfresx;
+            const double sinphi10 = sinphi0;
+            const double cosphi10 = cosphi0;
+            const double dn10sinphi10 = dn10 * sinphi10;
+            dX10 = -de10 * sinlam10 - dn10sinphi10 * coslam10;
+            dY10 = de10 * coslam10 - dn10sinphi10 * sinlam10;
+            dZ10 = dn10 * cosphi10;
+
+            const double sinlam11 = sinhalfresx;
+            const double coslam11 = coshalfresx;
+            const double sinphi11 = sinphi1;
+            const double cosphi11 = cosphi1;
+            const double dn11sinphi11 = dn11 * sinphi11;
+            dX11 = -de11 * sinlam11 - dn11sinphi11 * coslam11;
+            dY11 = de11 * coslam11 - dn11sinphi11 * sinlam11;
+            dZ11 = dn11 * cosphi11;
+        }
+
+        dX = m00 * dX00 + m01 * dX01 + m10 * dX10 + m11 * dX11;
+        dY = m00 * dY00 + m01 * dY01 + m10 * dY10 + m11 * dY11;
+        dZ = m00 * dZ00 + m01 * dZ01 + m10 * dZ10 + m11 * dZ11;
+    }
+};
+
+// ---------------------------------------------------------------------------
+
+enum class DisplacementType { NONE, HORIZONTAL, VERTICAL, THREE_D };
+
+// ---------------------------------------------------------------------------
+
+/** Internal class to offer caching services over a Component */
+template <class Grid, class GridSet> struct ComponentEx {
+    const Component &component;
+
+    const bool isBilinearInterpolation; /* bilinear vs geocentric_bilinear */
+
+    const DisplacementType displacementType;
+
+    // Cache
+    std::unique_ptr<GridSet> gridSet{};
+    std::map<const Grid *, GridEx<Grid>> mapGrids{};
+
+  private:
+    mutable double mCachedDt = 0;
+    mutable double mCachedValue = 0;
+
+    static DisplacementType getDisplacementType(const std::string &s) {
+        if (s == STR_HORIZONTAL)
+            return DisplacementType::HORIZONTAL;
+        if (s == STR_VERTICAL)
+            return DisplacementType::VERTICAL;
+        if (s == STR_3D)
+            return DisplacementType::THREE_D;
+        return DisplacementType::NONE;
+    }
+
+  public:
+    explicit ComponentEx(const Component &componentIn)
+        : component(componentIn),
+          isBilinearInterpolation(
+              componentIn.spatialModel().interpolationMethod == STR_BILINEAR),
+          displacementType(getDisplacementType(component.displacementType())) {}
+
+    double evaluateAt(double dt) const {
+        if (dt == mCachedDt)
+            return mCachedValue;
+        mCachedDt = dt;
+        mCachedValue = component.timeFunction()->evaluateAt(dt);
+        return mCachedValue;
+    }
+
+    void clearGridCache() {
+        gridSet.reset();
+        mapGrids.clear();
+    }
+};
+
+// ---------------------------------------------------------------------------
+
+/** Evaluator exception. */
+class EvaluatorException : public std::exception {
```

Comment:
These exceptions should be pushed into another header so this one isn't so long.

Reply:
done